Add DAG run telemetry metrics for load mode, invocation, and other render_config parameters

pankajkoti opened the pull request:
Adds 9 new telemetry metrics to DAG run events to better understand how Cosmos is configured and used:

  - `used_automatic_load_mode` - whether the user specified LoadMode.AUTOMATIC
  - `actual_load_mode` - the resolved load method (dbt_ls, dbt_manifest, custom, etc.)
  - `invocation_mode` - subprocess or dbt_runner
  - `install_deps` - whether dbt deps installation is enabled
  - `uses_node_converter` - whether custom node converters are used
  - `test_behavior` - how dbt tests are handled (after_each, after_all, none, etc.)
  - `source_behavior` - how dbt sources are rendered
  - `total_dbt_models` - total number of dbt models in project
  - `selected_dbt_models` - number of models selected for rendering
  
 These metrics are collected during DAG conversion and stored in `dag.params` to survive Airflow's serialisation process, then emitted by the DAG run listener on success/failure events.

closes: #2109

pankajkoti committed:
Add DAG run telemetry metadata collection

Store and emit 9 new Cosmos configuration metrics on DAG runs:
- used_automatic_load_mode: Whether LoadMode.AUTOMATIC was used
- actual_load_mode: The resolved load method (e.g., dbt_ls, dbt_manifest)
- invocation_mode: How dbt is invoked (subprocess, dbt_runner)
- install_deps: Whether dependency installation is enabled
- uses_node_converter: Whether custom node converters are used
- test_behavior: Test rendering behavior (after_each, none, etc.)
- source_behavior: Source rendering behavior (all, none, etc.)
- total_dbt_models: Total number of dbt models in the project
- selected_dbt_models: Number of models selected after filtering

Implementation:
- Added _store_cosmos_telemetry_metadata_on_dag() in converter to store metadata on DAG object
- Added get_cosmos_telemetry_metadata() helper in dag_run_listener to extract metadata
- Updated on_dag_run_success and on_dag_run_failed hooks to include metadata in telemetry
- Added comprehensive tests for both success and failure scenarios
- Fixed existing test that was affected by additional debug logging

pankajkoti committed:
Fix DAG run telemetry metadata serialization issue

The metadata was stored as a custom attribute (_cosmos_telemetry_metadata) which is not preserved during Airflow DAG serialization. When the dag_run_listener receives the DAG, it gets a SerializedDAG where custom attributes are lost, resulting in an empty metadata dictionary.

Solution: Store metadata in dag.params which is serialized by Airflow and accessible in the listener. Using key __cosmos_telemetry_metadata__ to avoid conflicts with user-defined params.

Changes:

- Store metadata in dag.params[__cosmos_telemetry_metadata__] in converter

- Retrieve from dag.params.get(__cosmos_telemetry_metadata__, {}) in listener

- Updated docstrings to reflect the new storage mechanism

pankajkoti committed:
Fix test validation error in dag_run_listener test

pankajkoti committed:
Ensure telemetry metadata values are JSON-serializable

pankajkoti committed:
Fix integration test to use manifest load mode

pankajkoti committed:
Add manifest_path to failed telemetry test

pankajkoti committed:
Add tests for telemetry metadata exception handlers

Covers exception handling in _store_cosmos_telemetry_metadata_on_dag
to ensure graceful degradation when metrics computation fails. Tests
verify that 8 out of 9 exception handlers work correctly for actual_load_mode,
invocation_mode, install_deps, uses_node_converter, test_behavior,
source_behavior, total_dbt_models, and selected_dbt_models.

tatiana commented on cosmos/converter.py:
```diff
@@ -370,3 +378,93 @@ def _add_dbt_project_hash_to_dag_docs(self, dag: DAG | None) -> None:
+
+        try:
+            metadata["actual_load_mode"] = str(self.dbt_graph.load_method.value)
+        except Exception as e:
```
Could we capture a more specific exception here? Which error do we expect to face?

pankajkoti replied:
yes, capturing specific exceptions now

tatiana commented on cosmos/converter.py:
```diff
@@ -370,3 +378,93 @@ def _add_dbt_project_hash_to_dag_docs(self, dag: DAG | None) -> None:
+            logger.warning(f"Failed to compute invocation_mode: {e}")
+
+        try:
+            install_deps = operator_args.get("install_deps")
```
Cosmos DAG parsing should not rely on `install_deps` from `operator_args`. It should rely exclusively in the `render_config` or `project_config` one

tatiana reviewed (commented):
Thanks for working on this, @pankajkoti. I'm glad the DAG params are working!

I left some feedback inline.

Please confirm which versions of Airflow you tested it against, and whether you were able to confirm the events reaching Scarf?

pankajkoti committed:
Address review comments

pankajkoti commented on tests/listeners/test_dag_run_listener.py:
```diff
@@ -183,3 +184,116 @@ def test_on_dag_run_failed(mock_emit_usage_metrics_if_enabled, caplog):
+@pytest.mark.skipif(
+    AIRFLOW_VERSION >= Version("3.1.0"),
+    reason="TODO: Fix create_dag_run to work with AF 3.1 and remove this skip.",
+)
```
```suggestion
```

pankajkoti commented on tests/listeners/test_dag_run_listener.py:
```diff
@@ -183,3 +184,116 @@ def test_on_dag_run_failed(mock_emit_usage_metrics_if_enabled, caplog):
+@pytest.mark.skipif(
+    AIRFLOW_VERSION >= Version("3.1.0"),
+    reason="TODO: Fix create_dag_run to work with AF 3.1 and remove this skip.",
+)
```
```suggestion
```

pankajkoti committed:
Apply suggestion from @pankajkoti

pankajkoti committed:
Apply suggestion from @pankajkoti

pankajkoti committed:
Merge branch 'main' into telemetry-dagrun-metrics

pankajkoti committed:
Fix tests

pankajkoti commented:
> Thanks for working on this, @pankajkoti. I'm glad the DAG params are working!
> 
> I left some feedback inline.
> 
> Please confirm which versions of Airflow you tested it against, and whether you were able to confirm the events reaching Scarf?

Yes, I am glad too that the DAG params worked here. Thanks for the review, it helped.

Addressed the review feedback. And I tested this with both Airflow 2 and 3, events are reaching Scarf on both, and we get 200 OK responses.

pankajkoti committed:
Address Copilot's review comment

pankajkoti commented on cosmos/converter.py:
```diff
@@ -33,6 +33,8 @@
 
 logger = get_logger(__name__)
 
+from cosmos.constants import DbtResourceType
```
yes, good catch, addressed.

tatiana commented on cosmos/converter.py:
```diff
@@ -294,6 +297,9 @@ def __init__(
 
         self._add_dbt_project_hash_to_dag_docs(dag)
+        self._store_cosmos_telemetry_metadata_on_dag(
+            dag, render_config, project_config, operator_args, initial_load_method
```
We should not need `operator_args` for the DAG render config 
```suggestion
            dag, render_config, project_config, initial_load_method
```

tatiana commented on cosmos/converter.py:
```diff
@@ -372,3 +378,87 @@ def _add_dbt_project_hash_to_dag_docs(self, dag: DAG | None) -> None:
+        :param dag: The Airflow DAG to store metadata on. If None, no action is taken.
+        :param render_config: The render configuration
+        :param project_config: The project configuration
+        :param operator_args: The operator arguments
```
```suggestion
```

tatiana commented on cosmos/converter.py:
```diff
@@ -372,3 +378,87 @@ def _add_dbt_project_hash_to_dag_docs(self, dag: DAG | None) -> None:
+        # Compute each metric individually with error handling
+        try:
+            metadata["used_automatic_load_mode"] = initial_load_method == LoadMode.AUTOMATIC
+        except (AttributeError, TypeError) as e:
```
In which scenario would we get `AttributeError` or `TypeError` here? We may not need any exception handling in this line

tatiana commented on cosmos/converter.py:
```diff
@@ -372,3 +378,87 @@ def _add_dbt_project_hash_to_dag_docs(self, dag: DAG | None) -> None:
+
+        try:
+            invocation_mode = None
+            if render_config.invocation_mode:
```
If `render_config` is not `None`, then it will have an `invocation_mode` attribute:
https://github.com/astronomer/astronomer-cosmos/blob/0fa0163c02d709e0849e007e40f165ef0743d29c/cosmos/config.py#L86

The same applies to `dbt_deps` below:
https://github.com/astronomer/astronomer-cosmos/blob/0fa0163c02d709e0849e007e40f165ef0743d29c/cosmos/config.py#L90

And also:
- `uses_node_converter`
- `test_behavior`
- `source_behavior`

Would it be worth simplifying the code by checking whether `render_config` is `None` and setting those values accordingly, without the need of so many `try`/`except` statements?

tatiana commented on cosmos/converter.py:
```diff
@@ -372,3 +378,87 @@ def _add_dbt_project_hash_to_dag_docs(self, dag: DAG | None) -> None:
+            logger.warning(f"Failed to compute used_automatic_load_mode: {e}")
+
+        try:
+            metadata["actual_load_mode"] = str(self.dbt_graph.load_method.value)
```
We may want to handle this close to the other dbt_graph properties:
- `total_dbt_models`
- `selected_dbt_models`

Applying the same feedback as described in https://github.com/astronomer/astronomer-cosmos/pull/2223/changes#r2655266426

tatiana reviewed (approved):
Thanks a lot, @pankajkoti , I left some minor feedback in line, but since you already tested and confirmed this worked with Airflow 2 and 3, I'm approving it - feel free to merge once the feedback is addressed.

pankajkoti committed:
Remove unused operator_args parameter from telemetry metadata method

The operator_args parameter in _store_cosmos_telemetry_metadata_on_dag was never used in the method body. All telemetry metadata is collected from render_config and project_config which are already passed as separate parameters.

Changes:
- Remove operator_args parameter from method signature and docstring
- Update method call site to remove the argument
- Remove unused mock_operator_args from test

pankajkoti committed:
Address review feedback on exception handling

pankajkoti committed:
Simplify test to match updated telemetry metadata implementation

pankajkoti committed:
Verify install_deps is captured from render_config during DAG parsing

pankajkoti committed:
Fix tests

pankajkoti commented on tests/listeners/test_dag_run_listener.py:
```diff
@@ -183,3 +184,116 @@ def test_on_dag_run_failed(mock_emit_usage_metrics_if_enabled, caplog):
+            test_behavior=TestBehavior.AFTER_EACH,
+            source_rendering_behavior=SourceRenderingBehavior.NONE,
+        ),
+        operator_args={"install_deps": True},
```
yes, I set it here to False. And set the render_config.dbt_deps=True and asserting that.

pankajkoti commented on tests/listeners/test_dag_run_listener.py:
```diff
@@ -183,3 +184,116 @@ def test_on_dag_run_failed(mock_emit_usage_metrics_if_enabled, caplog):
     assert mock_emit_usage_metrics_if_enabled.call_count == 1
+
+
+@pytest.mark.skipif(
```
created an issue for this: https://github.com/astronomer/astronomer-cosmos/issues/2244
